sql/analyzer/join_order_builder.go

```diff
@@ -394,12 +394,17 @@ func (j *joinOrderBuilder) addJoin(op plan.JoinType, s1, s2 vertexSet, joinFilte
 	if !ok {
 		group = j.memoize(op, left, right, joinFilter, selFilters)
 		j.plans[union] = group
-	} else {
-		j.addJoinToGroup(op, left, right, joinFilter, selFilters, group)
 	}
 	if commute(op) {
+		// Adding joins to the expression group places them at the start of the list, so if an operation is
+		// commutative, make sure we add the commuted version first, then add the original version, so that the original
+		// version is seen first. Otherwise, if both versions have the same, best cost, the analyzer will get stuck
+		// cycling between the two versions and the plan won't stabilize.
```

Comment:
If you look at the memo below in order_builder_test, you'll notice that the non-commuted versions are the canonical head of every memo group, and inserted for a second time before the commuted pair now. We shouldn't be inserting them the second time. But when we fix that bug, we will start thrashing plans again. I think the necessary solution in the future will default to the canonical plan in a cost tie, which by definition is the last member of the memo group.

Reply:
my first guess would be to have coster to pick that last valid plan it finds. also might be a good time to fix the bug that i made, by putting `last` or some form of `last` back as a canonical plan so we check for and avoid the plan duplication.

Reply:
Makes sense. Thanks for pointing that out; happy to tackle this, too. 

My first thought is that prepend (or a separate function) should check for equality with a `relExpr` already in the `exprGroup` and not add a duplicate, then when we select the best plan, we accept the plan that has a cost equal to, or less than the current best plan, which will still ensure that the canonical plan will win when there's a tie.